omnikinverter/tcp_model.py

```diff
@@ -0,0 +1,151 @@
+"""Data model and conversions for tcp-based communication."""
+from ctypes import BigEndianStructure, c_char, c_ubyte, c_uint, c_ushort
+from typing import Any
+
+from .exceptions import OmnikInverterPacketInvalidError
+
+UINT16_MAX = 65535
+
+
+class _AcOutput(BigEndianStructure):
+    _fields_ = [
+        ("frequency", c_ushort),
+        ("power", c_ushort),
+    ]
+
+    def get_power(self):
+        return None if self.power == UINT16_MAX else self.power
+
+    def get_frequency(self):
+        return None if self.frequency == UINT16_MAX else self.frequency * 0.01
+
+
+class _TcpData(BigEndianStructure):
+    _pack_ = 1
+    _fields_ = [
+        ("magic", c_char * 3),
+        ("padding0", c_ubyte * 11),
+        ("serial_number", c_char * 16),
+        ("temperature", c_ushort),
+        ("dc_input_voltage", c_ushort * 3),
+        ("dc_input_current", c_ushort * 3),
+        ("ac_output_current", c_ushort * 3),
+        ("ac_output_voltage", c_ushort * 3),
+        ("ac_output", _AcOutput * 3),
+        ("solar_energy_today", c_ushort),
+        ("solar_energy_total", c_uint),
+        ("solar_hours_total", c_uint),
+        ("padding1", c_ubyte * 6),
+        ("unknown0", c_ushort),
+        ("padding2", c_ubyte * 10),
+        ("firmware", c_char * 16),
+        ("padding4", c_uint),
+        ("firmware_slave", c_char * 16),
+        ("padding4", c_uint),
+    ]
+
+
+class TcpParser:
+    """Functions for interacting with the Omnik over TCP."""
```

Comment:
Since all of these are `@classmethod` anyway, perhaps these functions should just move into the module directly? Seems like an unnecessary indirection in hindsight.